shared/lib_battery_lifetime_lmolto.cpp

```diff
@@ -0,0 +1,183 @@
+or promote products derived from this software without specific prior written permission.
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED.IN NO EVENT SHALL THE COPYRIGHT HOLDER, CONTRIBUTORS, UNITED STATES GOVERNMENT OR UNITED STATES
+DEPARTMENT OF ENERGY, NOR ANY OF THEIR EMPLOYEES, BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+OR CONSEQUENTIAL DAMAGES(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <cmath>
+#include <numeric>
+
+#include "lib_battery_lifetime_lmolto.h"
+
+void lifetime_lmolto_t::initialize() {
+    state = std::make_shared<lifetime_state>();
+    // cycle model for counting cycles only, no cycle-only degradation
+    cycle_model = std::unique_ptr<lifetime_cycle_t>(new lifetime_cycle_t(params, state));
+    cycle_model->resetDailyCycles();
+    state->lmo_lto->dq_relative_cal = 0;
+    state->lmo_lto->dq_relative_cyc = 0;
+    state->lmo_lto->temp_avg = 0;
+    state->lmo_lto->EFC = 0;
+    state->lmo_lto->EFC_dt = 0;
+    state->q_relative = 100. - state->lmo_lto->dq_relative_cal - state->lmo_lto->dq_relative_cyc;
+}
+
+
+lifetime_lmolto_t::lifetime_lmolto_t(double dt_hr) {
+    params = std::make_shared<lifetime_params>();
+    params->model_choice = lifetime_params::LMOLTO;
+    params->dt_hr = dt_hr;
+    initialize();
+}
+
+lifetime_lmolto_t::lifetime_lmolto_t(std::shared_ptr<lifetime_params> params_pt) {
+    params = std::move(params_pt);
+    initialize();
+}
+
+lifetime_lmolto_t::lifetime_lmolto_t(std::shared_ptr<lifetime_params> params_pt, std::shared_ptr<lifetime_state> state_pt) {
+    params = std::move(params_pt);
+    state = std::move(state_pt);
+    cycle_model = std::unique_ptr<lifetime_cycle_t>(new lifetime_cycle_t(params, state));
+}
+
+lifetime_lmolto_t::lifetime_lmolto_t(const lifetime_lmolto_t &rhs) :
+        lifetime_t(rhs){
+    operator=(rhs);
+}
+
+lifetime_lmolto_t& lifetime_lmolto_t::operator=(const lifetime_lmolto_t& rhs) {
+    if (this != &rhs) {
+        *params = *rhs.params;
+        *state = *rhs.state;
+    }
+    return *this;
+}
+
+lifetime_t * lifetime_lmolto_t::clone() {
+    return new lifetime_lmolto_t(*this);
+}
+
+double lifetime_lmolto_t::runQcal() {
+    double SOC_avg = cycle_model->predictAvgSOC(state->cycle->DOD_max - state->cycle->DOD_min);
+    double q1 = q1_b0 * exp(q1_b1 * (1 / pow(state->lmo_lto->temp_avg, 3.)) * sqrt(SOC_avg)) *
+                      exp(q1_b2 * (1 / pow(state->lmo_lto->temp_avg, 2.)) * sqrt(SOC_avg));
```

Comment:
Are these variables going to be in the reference? Comments might be helpful.

Reply:
Yes, they will be in the reference. They are parameters to the equations, describing the effect of each term in the equation, so not too much to explain beyond that, but I added some comments about the equations